lifeomic_logging/logger.py

```diff
@@ -0,0 +1,235 @@
+import socket
+import os
+from datetime import datetime
+import threading
+import uuid
+from collections import OrderedDict
+from logging import Formatter, Filter, StreamHandler, getLogger, INFO, Logger
+from json import dumps
+from contextlib import contextmanager
+
+thread_local_storage = threading.local()
+
+_RESERVED_ATTRS = (
+    "args",
+    "asctime",
+    "created",
+    "exc_info",
+    "exc_text",
+    "filename",
+    "funcName",
+    "levelname",
+    "levelno",
+    "lineno",
+    "module",
+    "msecs",
+    "message",
+    "msg",
+    "name",
+    "pathname",
+    "process",
+    "processName",
+    "relativeCreated",
+    "stack_info",
+    "thread",
+    "threadName",
+)
+
+
+@contextmanager
+def scoped_logger(
+    logname: str,
+    normal_context: dict = None,
+    error_context: dict = None,
+    stream=None,
+    level: int = INFO,
+) -> Logger:
+    """Creates a scopped logger
+
+    Parameters
+    ----------
+    logname : str
+        The name of the logger
+    normal_context : dict, optional
+        Dict of values to add to the logger, by default None
+    error_context : dict, optional
+        Dict of values to add to the logger when an error occurs,
+        by default None
+    stream : Stream, optional
+        A stream to send the log events to, by default None
+    level : int, optional
+        The logging level, by default INFO
+
+    Returns
+    -------
+    logging.Logger
+        The scoped logger
+    """
+
+    try:
+        _LoggingContextFilter.push_context(
+            normal_context, _LoggingContextFilter.tls_normal_context_field_name
+        )
+        _LoggingContextFilter.push_context(
+            error_context, _LoggingContextFilter.tls_error_context_field_name
+        )
+
+        yield _logger(logname, StreamHandler(stream), level)
+    finally:
+        _LoggingContextFilter.pop_context(
+            _LoggingContextFilter.tls_normal_context_field_name
+        )
+        _LoggingContextFilter.pop_context(
+            _LoggingContextFilter.tls_error_context_field_name
+        )
+
+
+def get_request_context(lambda_context: dict) -> dict:
+    """Helper method to retrieve our most pertinent header information.
+        These are stored from the passed in lambda_context
+        object. Please see
+        https://docs.aws.amazon.com/lambda/latest/dg/python-logging.html
+        for the structure.
+
+    Parameters
+    ----------
+    lambda_context : dict
+        The lambda context
+
+    Returns
+    -------
+    dict
+        A dict of important properties for the request
+    """
+    if lambda_context is None:
+        return None
+
+    client_context = getattr(lambda_context, "client_context", None)
+    if not client_context:
+        return None
+
+    custom = getattr(client_context, "custom", {})
```

Comment:
Minor: let's default to `None`, don't need to create a dict instance we won't use